lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php

```diff
@@ -218,6 +218,30 @@ public function setTemplateOptions($templateOptions)
     }
 
     /**
+     * Creates a Zend_Mime_Part attachment and attaches it to the current message scope.
+     * Attachment is automatically added to the mail object after creation.
+     *
+     * See also: http://framework.zend.com/manual/1.12/de/zend.mail.attachments.html
+     *
+     * @param  string $body
+     * @param  string $mimeType
+     * @param  string $disposition
+     * @param  string $encoding
+     * @param  string $filename A filename for the attachment
+     * @return $this
+     */
+    public function addAttachment(
+        $body,
+        $mimeType = \Zend_Mime::TYPE_OCTETSTREAM,
+        $disposition = \Zend_Mime::DISPOSITION_ATTACHMENT,
+        $encoding = \Zend_Mime::ENCODING_BASE64,
+        $filename = null
+    ) {
+        $this->message->createAttachment($body, $mimeType, $disposition, $encoding, $filename);
```

Comment:
Indeed, there are a couple of ways to tackle this.

One Java-ish method would be to create a new `MessageInterface2`, which extends the original. There you can add any method you want, as it's not used by anyone. In due time, the original would be deprecated, then removed in favour of the new one.

Another way would be the Decorator pattern (`MessageWithAttachment` or something).

Either way, you would still have the problem of leaking implementation details.
You'll need to put more thought in this and create a new `MessageAttachment` interface which should satisfy most of the popular frameworks out there. From what I can see, it usually boils down to filename, encoding, content type and disposition. This is the case with `\Zend_Mime` _and_ `Symfony`'s own builder: http://swiftmailer.org/docs/messages.html#attaching-files. This functionality lends itself perfectly for abstraction. So do create such an interface with methods like `getDisposition()`, `getFileName()` etc., then you can create a `Zend_Mime`-compatible implementation and set a preference to that in M2's DI. Most use SwiftMailer anyway, so if you consider that, you're covering most of the userbase.